Add multi-node administration section (#651)
timescale/docs: timescaledb/how-to-guides/multinode-timescaledb/multinode-administration.md

* Schemas in distributed databases
* Default privileges in distributed databases
* Distributed hypertables

When you perform an administrative task, such as adding a new
schema, changing a default privilege, or adding an index on a
distributed hypertable, you can perform the task from the access node
and it is applied to all data nodes in the distributed database. If a
command is executed on a regular table, however, the effects of that
command are only applied locally on the access node. Similarly, if a
command is executed directly on a data node, the result is only
visible on that data node.

There are some exceptions to this rule. Commands that act on the
distributed database itself, such as [`ALTER
DATABASE`][alter-database], are not distributed to data nodes. That is
because these settings usually need to be different on the access node
compared to the data nodes. For example, the publicly accessible
access node typically requires a different connection limit than the
data nodes that are only accessed internally. The nodes could also be
provisioned differently, having varying levels of CPU, memory, and
disk capabilities, so the database-level configuration must be

> heterogeneous and reflect the node capabilities.

Reviewer: and "should" reflect

Reviewer 2: ```suggestion
heterogeneous and should reflect the node capabilities.
```

use the [`distributed_exec`][distributed_exec] procedure to create and
configure roles across all data nodes in the current database.

### Why roles are not managed by TimescaleDB
In PostgreSQL, most objects, such as tables, views, and indexes, live
within a particular database. In a distributed database, this also
includes the metadata about a multi-node cluster, such as the data
nodes the access node knows about. In contrast, roles are *shared*
objects that exist across all databases, distributed and
non-distributed alike. Thus, when creating a role from a
non-distributed database there is no information in the database on
how to distribute the command. Further, different distributed
databases on the same access node can have a different set of data
nodes. For example, in a multi-tenancy system, one distributed
database might only be configured with a subset of data nodes because
one wants to limit the usage of the data nodes based on the database
owner. Therefore, distributing role management commands based on the
data nodes in one distributed database might not work for all the
distributed databases on the same access node.

### Creating a distributed role

> When creating a distributed role, it is important to note that the

Reviewer 2: ```suggestion
When you create a distributed role, it is important to consider that the
```

same role might require different configuration on the access node
compared to the data nodes. For example, a user might require a
password to connect to the access node, while certificate

> authentication is used between nodes within the cluster. One might

Reviewer 2: ```suggestion
authentication is used between nodes within the cluster. You might
```

> also want a connection limit for external connections but allow

Reviewer 2: ```suggestion
also want a connection limit for external connections, but allow
```

> unlimited connections to data nodes as every connection to an access

Reviewer 2: ```suggestion
unlimited connections to data nodes. For example, every connection to an access
```

node could potentially spawn multiple connections to the same data
node in the backend:

> 

Reviewer 2: ```suggestion
```

```sql
CREATE ROLE alice WITH LOGIN PASSWORD 'mypassword' CONNECTION LIMIT 10;
CALL distributed_exec($$ CREATE ROLE alice WITH LOGIN CONNECTION LIMIT -1; $$);
```

> Note that certificate authentication is configured separately from the
> roles themselves. See the section on [multi-node
> authentication][multi-node-authentication] for more information.

Reviewer 2: ```suggestion
Certificate authentication is configured separately from the
roles. For more information about authentication, see the 
[multi-node authentication section][multi-node-authentication].
```

> Some roles might also be configured without the `LOGIN` attribute on
> the access node so that one can switch to the role locally but not
> connect with the user from a remote location. However, to be able to
> connect from the access node to a data node as that user, the data
> nodes need to have the role configured with the `LOGIN` attribute
> enabled. To create a non-login role for a multi-node setup, issue the
> following commands:
> ```sql
> CREATE ROLE alice WITHOUT LOGIN;
> CALL distributed_exec($$ CREATE ROLE alice WITH LOGIN; $$);
> ```

Reviewer 2: ```suggestion
Some roles can also be configured without the `LOGIN` attribute on
the access node. This allows you to switch to the role locally, but not
connect with the user from a remote location. However, to be able to
connect from the access node to a data node as that user, the data
nodes need to have the role configured with the `LOGIN` attribute
enabled. To create a non-login role for a multi-node setup, use these
commands:
```sql
CREATE ROLE alice WITHOUT LOGIN;
CALL distributed_exec($$ CREATE ROLE alice WITH LOGIN; $$);
```
```

> To allow a new role to create distributed hypertables it also needs to
> be granted usage on data nodes:

Reviewer 2: ```suggestion
To allow a new role to create distributed hypertables it also needs to
be granted usage on data nodes, for example:
```

```sql
GRANT USAGE ON FOREIGN SERVER dn1,dn2,dn3 TO alice;
```

> By granting usage on some data nodes, but not others, it is possible
> to restrict usage to a subset of data nodes based on the role.

Reviewer 2: ```suggestion
By granting usage on some data nodes, but not others, you can
restrict usage to a subset of data nodes based on the role.
```

> ### Altering a distributed role

Reviewer 2: ```suggestion
### Alter a distributed role
```

> When altering a distributed role, the same process is followed as when
> creating roles. The role needs to altered locally and on the data
> nodes in two separate steps. For example, to add the `CREATEROLE`
> attribute to a role do:
> 

Reviewer 2: ```suggestion
When you alter a distributed role, use the same process as 
creating roles. The role needs to be altered locally and on the data
nodes in two separate steps. For example, to add the `CREATEROLE`
attribute to a role:
```

```sql
ALTER ROLE alice CREATEROLE;
CALL distributed_exec($$ ALTER ROLE alice CREATEROLE; $$);
```

> ## Managing distributed databases

Reviewer 2: ```suggestion
## Manage distributed databases
```

A distributed database can contain both distributed and
non-distributed objects. In general, when a command is issued to alter
a distributed object it applies to all nodes that have that object (or
a part of it).

However, in some cases settings *should* be different depending on
node, because nodes might be provisioned differently (having, e.g.,
varying levels of CPU, memory, and disk capabilities) and the role of
the access node is different from a data node's.

This section describes how and when commands on distributed objects
are applied across all data nodes when executed from within a

> distributed database.

Reviewer 2: Dupe?

> ### Altering a distributed database

Reviewer 2: ```suggestion
### Alter a distributed database
```

> An [`ALTER DATABASE`][alter-database] command is only applied locally on the access
> node. There are several reasons why TimescaleDB does *not* apply this
> command across all data nodes.
> 
> First, a distributed database supports having different database names
> on each data node. Therefore, renaming a distributed database on the
> access node is a local operation that applies only to the access node
> database. Note that when renaming a data node database the
> configuration of the data node must also be updated on the access
> node.
> 

Reviewer 2: ```suggestion
The [`ALTER DATABASE`][alter-database] command is only applied locally on the access
node. This is because a distributed database supports having different database names
on each data node. Therefore, renaming a distributed database on the
access node is a local operation that applies only to the access node
database. When you rename a data node database, the
configuration of the data node must also be updated on the access
node.
```

> Second, database-level configuration parameters should often be
> different across nodes. For example:
> 
> ```sql
> ALTER DATABASE mydatabase SET max_parallel_workers TO 12;
> ```
> 
> is a setting that might differ depending on the CPU capabilities of
> the node.
> 
> Third, a database option like:
> 
> ```sql
> ALTER DATABASE mydatabase WITH CONNECTION_LIMIT 10;
> ```
> 
> is also, in most cases, different on an access node compared to its
> data nodes.

Reviewer 2: ```suggestion
Additionally, database-level configuration parameters often need to be
different across nodes. For example, this is a setting that might differ 
depending on the CPU capabilities of the node:
```sql
ALTER DATABASE mydatabase SET max_parallel_workers TO 12;
```

It is also possible that a database option like is different on an access node compared to its
data nodes:
```sql
ALTER DATABASE mydatabase WITH CONNECTION_LIMIT 10;
```

```

> ### Dropping a distributed database

Reviewer 2: ```suggestion
### Drop a distributed database
```

> When issuing the command to drop a distributed database on the access
> node, it will *not* automatically drop the corresponding databases on
> the data nodes. Instead, it is necessary to connect directly to each
> data node and drop the databases locally.

Reviewer 2: ```suggestion
When you drop a distributed database on the access node, it does not 
automatically drop the corresponding databases on the data nodes. 
In this case, you need to connect directly to each data node and drop 
the databases locally. 
```

> There are several reasons for why a distributed database is not
> automatically dropped across all nodes. 
> 
> First, the information about data nodes lives within the distributed
> database on the access node, but it is not possible to read it when
> executing the drop command since it cannot be issued when connected to
> the database.
> 
> Second, in case a data node has permanently failed, one wants to be
> able to drop a database even if one or more data nodes are not
> responding.
> 
> Third, it is often good to leave the data intact on a data node out of
> precaution. For example, one might want to backup a data node even
> after a database was dropped on the access node.

Reviewer 2: ```suggestion
A distributed database is not automatically dropped across all nodes, 
because the information about data nodes lives within the distributed 
database on the access node, but it is not possible to read it when 
executing the drop command since it cannot be issued when 
connected to the database.

Additionally, if a data node has permanently failed, you need to be able 
to drop a database even if one or more data nodes are not responding.

It is also good practice to leave the data intact on a data node if 
possible. For example, you might want to back up a data node even 
after a database was dropped on the access node.
```

> ## Creating, altering, and dropping schemas

Reviewer 2: ```suggestion
## Create, alter, and drop schemas
```

> When creating, altering, or dropping schemas, the commands are applied
> across all data nodes in the current distributed database. Note,
> however, that existing schemas on the access node are not created on a
> data node when it is added. A missing schema is, however, created when
> a distributed hypertable is created and the schema it belongs to does
> not exist on a data node.
> 
> If a schema is created with a particular authorization, then the
> authorized role must also exist on the data nodes prior to issuing the
> command. The same things applies to altering the owner of an existing
> schema.

Reviewer 2: ```suggestion
When you create, alter, or drop schemas, the commands are applied
across all data nodes in the current distributed database. However, 
existing schemas on the access node are not created on a data node 
when it is added. A missing schema is, however, created when a 
distributed hypertable is created, and the schema it belongs to does
not exist on a data node.

If a schema is created with a particular authorization, then the
authorized role must also exist on the data nodes prior to issuing the
command. The same things applies to altering the owner of an existing
schema.
```

> ### Dropping owned objects
> The command [`DROP OWNED`][drop-owned] can be used to drop all objects
> owned by a role in order to prepare the role for removal. When
> executed, the command is applied across all data nodes of a
> distributed database.

Reviewer 2: ```suggestion
### Drop owned objects
Use the [`DROP OWNED`][drop-owned] command to drop all objects
owned by a role and prepare the role for removal. This command is 
applied across all data nodes of a distributed database.
```

> ### Managing privileges
> Privileges configured using [`GRANT`][grant] or [`REVOKE`][revoke]
> statements are applied to all data nodes if executed on a distributed
> object, like a distributed hypertable, schema in a distributed
> database, or the distributed database itself.
> 
> If the statement is executed on an object that a distributed object
> depends on, such as a sequence or index that is used by a distributed
> hypertable, the command is currently *not* applied across all data
> nodes.

Reviewer 2: ```suggestion
### Manage privileges
Privileges configured using [`GRANT`][grant] or [`REVOKE`][revoke]
statements are applied to all data nodes when they are run on a 
distributed object, like a distributed hypertable, schema in a distributed
database, or the distributed database itself.

If the statement is run on an object that a distributed object
depends on, such as a sequence or index that is used by a distributed
hypertable, the command is not applied across all data nodes.
```

> #### Setting default privileges
> Changes to default privileges are applied across all data nodes of a
> distributed database. The roles and schemas that the default
> privileges reference need to exist on the data nodes prior to
> executing the command.
> 
> Note that new data nodes are assumed to already have any altered
> default privileges. The default privileges are not automatically
> applied retrospectively to new data nodes.

Reviewer 2: ```suggestion
#### Set default privileges
Changes to default privileges are applied across all data nodes of a
distributed database. The roles and schemas that the default
privileges reference need to exist on the data nodes prior to
executing the command.

New data nodes are assumed to already have any altered
default privileges. The default privileges are not automatically
applied retrospectively to new data nodes.
```

retroactively?

> ## Managing tablespaces
> Tablespaces are instance-level objects like roles and therefore
> require manual handling on each node.

Reviewer 2: ```suggestion
## Manage tablespaces
Tablespaces are instance-level objects, like roles, and require 
manual configuration on each node.
```
